README.md
# IRA: 680x0 Interactive ReAssembler

IRA is a CLI _ReAssembler_ which translates any 68xxxx machine code executable or
binary file into an assembler sourcecode that might immediately be translated
back by an assembler.

                      ________________
                     /___/_________/_/\----- - -  -   -   -
                    /   /         /  \ \
                   /   /  ____   /    \ \
                  /   /  / / /  /  /\  \ \----- - -  -   -   -
                 /   /  / / /  /  / /\  \ \
                 \   \  \ \/  /  / /  \  \ \
                  \   \  \/  /  /  ¯¯¯¯\  \ \----- - -  -  -   -
                   \   \     \  ¯¯¯¯¯¯¯¯   \ \
                    \   \     \    ______   \ \ :
                    /   /  /\  \   \ \  :\   \ \|/--- - -  -  -   -
                   /   /  / /\  \   \ \ ° \   \ *--
                  /   /  / /  \  \   \/    \   \|\
                  ¯¯¯¯¯¯¯¯¯    ¯¯¯¯¯¯¯      ¯¯¯¯:



## Usage

    $ ./ira [OPTIONS] <SOURCE> [TARGET]


* Refer to [ira.doc](ira.doc) for the original IRA V1.xx documentation.
* Refer to [ira2.doc](ira2.doc) for IRA V2.xx documentation.
* Refer to [ira_config.doc](ira_config.doc) for a description of config directives.

### Example

Using the following commands to reassemble AmigaBASIC and to assemble an
identical binary with `vasm`. The config file was manually adjusted to reflect
all valid code regions (create a config file with `-preproc` first), then
duplicated as `NewAmigaBASIC.cnf`. The diff test reports no differences!

    $ ira -a -compat=bi -config -keepzh AmigaBASIC
    $ vasmm68k_mot -no-opt -Fhunkexe -nosym -o NewAmigaBASIC AmigaBASIC.asm
    $ ira -a -compat=bi -config -keepzh NewAmigaBASIC
    $ diff -s AmigaBASIC.asm NewAmigaBASIC.asm

The option `-compat=bi` is needed to allow bad btst instructions which access a
bit number > 7 in a byte (b) and to recognize immediate byte addressing modes
with an MSB of 0xff (i), which both appear frequently in the program. The
`-keepzh` option preserves empty sections, so that the number of sections stays
the same as before.

### Command-line args (default values are in brackets)

```
-M68xxx (-M68000)
        This specifies the type of processor or coprocessor for which the
        program should be reassembled. This doesn't have anything to do with
        the machine you're running IRA on.
        Most programs are written for 68000 CPUs.
        If anything other than -M68000 is specified, a MACHINE directive
        in the .ASM file is created. This works fine with PhxAss and vasm
        assemblers. For any other assembler this directive possibly has to be
        changed.

        Values accepted for xxx can be:

        - processor
          000,010,020,030,040,060
        - coprocessor
          851,881,882

-BINARY (off)
        IRA automatically recognises if the sourcefile is an executable,
        an object file or any other (binary). It may happen that some
        kind of binary data is recognised as an executable. To avoid this
        make use of the -BINARY option.

-A      (off)
        This option makes IRA append the address and data of an instruction
        to every line. That is pretty useful to me. E.g. when code and data
        is mixed, you can manually delete some instructions to replace them
        by DC.x directives.

-AW (off)
        Same as -A, but enforce 8-digit addresses.

-INFO   (off)
        Use this option to get some information about the hunk structure of
        your sourcefile.

-OFFSET=<OFFSET> (-OFFSET=0)
        When the sourcefile is relocated by IRA the offset value is added to
        the relocation. Why should I do that ?
        If you want to run an executable at a specific location in memory
        you specify the address of that location with the -OFFSET option
        in combination with the -KEEPBIN option. After running IRA you can
        load the .BIN file to the specific location and execute it. You just
        have to know what you're doing (e.g. contents of registers and so on).
        The second, more useful application is to write a part of memory to a
        file, then create a .ASM file with the -OFFSET option. You have to
        take the address of the memory location as offset. E.g. you can create
        your own kickfile with your own modifications (is it legal ?), of
        course some additional work on the .ASM file has to be done.
        OFFSET can be decimal or hexadecimal (e.g. -OFFSET=$4FF0).

-TEXT=1 (off)
        Since version 1.03 there is only one method for finding text left. The
        whole code and data sections are searched for text which is printed
        out to <stdout>. It may happen that some stupid text lines show up, so
        you have to decide yourself about every line to include into the .ASM
        file.

-KEEPZH (off)
        There are files that have hunks with a length of zero. By default
        these hunks don't appear in the .ASM file. If you want them preserved
        just use this option. There are executables where you have to use
        this option, because they work with their own SEGMENT structure.

-KEEPBIN (off)
        Before the first pass, an executable is relocated by IRA and written
        to a .BIN file. Normally, this file is deleted at some point, but if
        you want to keep it for some purpose, use the -KEEPBIN option.
        E.g. for a <type >x.hex x.bin opt h>.

-OLDSTYLE (depends on the -M68xxx option)
        This option forces IRA to use the old Motorola syntax like D16(PC)
        instead of (D16,PC). By default this option is used for 68000 and
        68010 processors.

-NEWSTYLE (depends on the -M68xxx option)
        This option forces IRA to use the new Motorola syntax like (D16,PC)
        instead of D16(PC). By default this option is used for 68020, 68030,
        68040 and 68060 processors.

-ESCCODES (off)
        Use escape character '\\' in strings.

-COMPAT=x (off)
        Various compatibility flags.

        b : Recognize immediate values of 8-15 for bit-instructions accessing
            memory (former -BITRANGE option).
        i : Recognize immediate byte addressing modes with an MSB of 0xff.
            Some assemblers generated 0xffff instead of 0x00ff for #-1.

-SPLITFILE (off)
        With this option the .ASM file is split up. Every section is put into
        its own file. One 'main' file is created to include these files via
        INCLUDE directives. What is it for ? I don't know. I was asked for
        this option.

-CONFIG (off)
        First, you can control the IRA settings with this option. That means
        you can specify the parameters that are otherways controlled via the
        command-line options. Second, you can specify addresses where to find
        code. This is useful for addresses where the -PREPROC option oversees
        code (for whatever reason). Third, you can specify symbols that are
        inserted instead using the 'LAB_xxxx' type.

        In combination with the -PREPROC option a new .cnf file is created.
        All the code areas found by PASS 0 are then written to this .cnf file.
        In addition, most of the command-line options are included, too.
        So, for the next calling of IRA you won't need to specify all of the
        command-line options you used to.

        If a .cnf file exists, a possible new one does not overwrite the old
        one. The name of the new one just get a '1' appended to its name.
        Look out for that.

        See ira_config.doc for details.

        Try the -CONFIG option out in combination with the -PREPROC option.
        IRA will create a .cnf file that you can look at for better
        understanding.

-PREPROC (off)
        This option will turn on PASS 0. That is to find data in code sections
        and even code in data sections (e.g some old compilers put their
        jumptables into data sections). Found data is checked for text.
        This option normally works very fine for compiled programs. But there
        may be problems like these:

        - parts of code may be seen as data. This comes for
          o code that is jumped to by (An), D16(An) or D8(PC,An)
            (jumptables, pointers to code, ...).
          o interrupt code that is only referenced by pointer (installation).
          o code that is never used.

        - parts of data may be seen as code. This comes for
          o crypted or crunched code.

-ENTRY=<OFFSET> (-ENTRY=0)
        If you know, that a file has data at the beginning (eg. bootblocks),
        you specify the (relative) adress where IRA should begin with
        code-scanning. For bootblocks: -ENTRY=$C .

-BASEREG[=n[,adr,sec]]
        n is the number of the base register, adr the address with that the
        base register is loaded and sec the section that n is related to.

        You can use this option if the program uses the smalldata model. It
        will provide you with a more readable .ASM file.
        Smalldata model means the access to data is made by D16(An), and the
        register An has to be preloaded by the SMALLDATABASE value. A lot of
        compilers use A4 as baseregister and the address of the datasection
        plus 32766 as the SMALLDATABASE.
        A good way to find out if a program makes use of the smalldata model
        is the following:
        1. Type "IRA >x -a -info -basereg test"
        2. Look at the file test.asm and x with an editor.
           If there are memory accesses by D16(An) (e.g. move.l -32754(A4),D0)
           look at the file x. There may be lines like BASEREG 00000008: A4.
           The first number is the hexadecimal address of an instruction that
           has A4 as destination register. Look at this address in the file
           test.asm. You may find a line like LEA SECSTRT_1,A4. Perhaps, A4
           is the base register and SECSTRT_1 the SMALLDATABASE. Now memo the
           address of SECSTRT_1 (=adr).
        3. Type "IRA -a -info -basereg=4,adr,1 test"
        4. Look at test.asm and you will see LAB_xxxx instead of D16(A4).
        5. The line NEAR A4,1 in test.asm tells the assembler to use the
           smalldata model. This directive may differ from assembler to
           assembler.

        As always, be careful when modifying a program. Often code and data
        is mixed or there are some program protection techniques that makes
        it hard to modify and run a program.

-BASEABS (off)
        When specified, a label in base-relative addressing mode is written as
        an absolute label, without the base register name (as with IRA V1.05).
        The default behaviour now is to write base-relative references
        as "(label,An)".
```



## Installation


CHANGELOG.md
# IRA changelog


## 2.09

### Bug fixes

- Fixed possible crash with low memory.
- Fixed CINV and CPUSH (only "LINE" operations were recognized).
- Fixed input file names with deep paths.
- Fixed negative values on 64-bit systems.
- Fixed MOVE16 (only MOVE16 (Ax)+,(Ay)+ were recognized).
- Fixed MOVEC (incorrect control register).
- Fixed MOVES (extension word were ignored).
- Fixed BFINS (swapped operands).

### New features

- Support for extended memory attributes in Amiga executables.
- Improve CPU choice on command line.
- New general documentation file (ira2.doc).
- New documentation file about config file's directives (ira_config.doc).
- New directives for config file: COMMENTS, BANNER, EQU and LABEL (see
  ira_config.doc).
- Improve error messages.
- Also support '0x' to denote hexadecimal addresses on the command line.
- New option -AW enforces 8-digit address output, when printing address
  and data in the comment field.
- Support for 68060 instructions (except FPU).
- Support for MMU instructions.


## 2.08

## Bug fixes

- Fixed problems with printf output formats for 64-bit compilers.
- $083c is illegal and was erroneously disassembled to BTST #n,CCR.
- Fixed CMP2, CHK2, CAS, CAS2.

## New features

- Do not overwrite an existing config file, when -preproc is specified.
- Make it work on Mac OSX 64-bit systems.


## 2.07

## Bug fixes

- A lot more fixes to make IRA clean for 64-bit compilers. Seems to work
  much better now.
- A base address of 0 should be allowed with binary files.

## New features

- Allow comments ';' and empty lines in the config file.


## 2.06

## Bug fixes

- Make IRA compile and work on 64-bit systems (using a 64-bit compiler).
- An empty register list is now written as #0 instead of (NOREG!). This
  makes it assemble (with vasm at least).
- Fixed label-generation for references into the middle of a data relocation
  (e.g. use "label+2" instead of making a new label, which caused trouble).
- Put both sides of the subtract operation into parentheses, when generating
  code for a jump table (to avoid problems with "label1-label2+2").
- Fixed some possible crashes.


## 2.05

## Bug fixes

- Fixed a possible crash when a config file defines a SYMBOL where the
  assigned value is not within the reassembled address range.
- Fixed a crash on encountering relocations at an odd address. This cannot
  easily be supported at the moment, so IRA will just quit when it happens.
- Reverted some stupid changes of the last release. Always prefer a label
  from the referring section in case both have the same address.

## New features

- Config file directive NOPTRS. Syntax is like PTRS, but it will prevent IRA
  from taking any address in this region as a program pointer (which would
  generate a label). Only works with binary input files!


## 2.04

## Bug fixes

- Fixed automatic ROM tag detection routine. Handling of absolute 32-bit
  function tables did not work when a relative 16-bit one was found before.
- Function names from ROM tag function tables were sometimes missing, when
  the input file is a raw binary.
- Fixed SECSTRT_n symbol recognition with BASEREG directive and small data
  addressing modes.
- Do not forget .W extension for word-sized LINK, MOVEA and MOVEP.
- Recognize BTST Dn,#x.

## New features

- Report about misplaced relocations in code (usually caused by a bad PTRS
  directive).
- Print a warning when a symbol from HUNK_SYMBOL is not inside the current
  section limits, and ignore this symbol.
- Base-relative symbols outside the small-data section's bounds are
  referenced via SECTSTRT_n and a warning is printed (because the
  instruction could be data, or the base register contains something
  else at this point).


## 2.03

## Bug fixes

- Make sure that CODE areas, when read from a config file, are split at
  section boundaries. Otherwise IRA cannot detect the start of a new section
  during source generation.
- BASEADR didn't work correctly for raw binary files.

## New features

- Config file directive TEXT. To define a region in data as printable text.
  This overrides the automatic text recognition.
  Syntax: `TEXT $<start> - $<end>`.
- Config file directives JMPB, JMPW and JMPL for generating jump-tables
  (or other offset-tables used to reference a program address).
  Syntax: `JMP<s> $<start> - $<end> [@ $<base>]`. `<s>` may be B, W or L and
  defines the width of the table entries (8, 16, 32 bit). The `<base>` is
  optional and same as `<start>`, when missing. It defines the base address
  where the table-offsets are added to.
- Option -BITRANGE, which was introduced in V2.00, was replaced by
  `-COMPAT=<flags>` to allow multiple compatibility flags. Currently known:
  * `b` : Recognize immediate values of 8-15 for bit-instructions accessing
          memory (former -BITRANGE option).
  * `i` : Recognize immediate byte addressing modes with an MSB of 0xff. Some
          assemblers generated 0xffff instead of 0x00ff for #-1.


## 2.02

## Bug fixes

- Fixed NBAS config directive. Only the entry with the highest address worked.
- File buffers for binary and config file name were too small. Extended from
  32 to 128 bytes (as source and target file name buffer already were).
- An immediate width-field in bitfield instructions showed up as '0' when
  it should have been '32'.
- Multiple MACHINE directives in the config file confused IRA.

init.c
/*
 * init.c
 *
 *  Created on: 24 december 2014
 *      Author   : Nicolas Bastien
 *      Project  : IRA  -  680x0 Interactive ReAssembler
 *      Part     : init.c
 *      Purpose  : IRA initialization subroutines
 *      Copyright: (C)2014-2016 Nicolas Bastien
 */

#include <ctype.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "ira.h"

#include "amiga_hunks.h"
#include "atari.h"
#include "binary.h"
#include "elf.h"
#include "init.h"
#include "ira_2.h"
#include "config.h"
#include "constants.h"
#include "supp.h"

ira_t *Start(void) {
    ira_t *ira;
    int i;

    /* note: mycalloc doesn't return if allocation failed, so no need to check returned value */
    ira = (ira_t *) mycalloc(sizeof(ira_t));

    ira->params.cpuType = NOCPU;
    ira->params.baseReg = 4;

    ira->reloc.relocMax = 1024;
    ira->symbols.symbolMax = 16;
    ira->codeArea.codeAreaMax = 16;
    ira->codeArea.cnfCodeAreaMax = 16;
    ira->codeArea.codeAdrMax = 16;
    ira->baseReg.baseSection = -1;
    ira->noBase.noBaseMax = 16;
    ira->noPtr.noPtrMax = 16;
    ira->text.textMax = 16;
    ira->jmp.jmpMax = 16;
    ira->label.labelMax = 1024;

    ira->pass = -1;
    ira->LabX_len = 400;
    ira->adrlen = 0;

    return ira;
}

void Init(ira_t *ira, int argc, char **argv) {
    int nextarg = 1;
    uint32_t i;
    uint16_t addrstyle = CPU_ADDR_STYLE;

    /* If argc lesser than 2: IRA is missing some arguments */
    if (argc < 2) {
#ifdef AMIGAOS
        /* argc equals 0 means "launched from Workbench" on AmigaOS. */
        if (argc == 0)
            fprintf(stderr, "IRA launching from Workbench is not yet supported\n\n");
#endif
        FormatError();
        ExitPrg("");
    }

    fprintf(stderr, IDSTRING1, VERSION, REVISION);

    /* Get everything we can from command line */
    ReadOptions(ira, argc, argv, &nextarg, &addrstyle);

    /* Let's check CPU and FPUs */
    CheckCPU(ira);

    /* By default, CPU will choose the addressing style (old style for 68000/010, new style for others).
     * If a style is explicitly asked, it will be used ignoring CPU. */
    if ((addrstyle == CPU_ADDR_STYLE && (ira->params.cpuType & (M68000 | M68010))) || addrstyle == OLD_ADDR_STYLE)
        ira->params.pFlags |= OLDSTYLE;

    /* All options are now handled, let's go to source and target file's name stuff. */

    /* Is there anymore argument that would contain the source file name ? */
    if (nextarg >= argc)
        ExitPrg("No source specified!");

    /* note: mystrdup() uses myalloc() which doesn't return if allocation failed, so no need to check returned value */
    ira->filenames.sourceName = mystrdup(argv[nextarg++]);

    /* Let's check if source file is readable. */
    if (!(ira->files.sourceFile = fopen(ira->filenames.sourceName, "rb")))
        ExitPrg("Can't open source file \"%s\".", ira->filenames.sourceName);

    /* If target is specified (in next argument), so be it */
    if (nextarg < argc)
        ira->filenames.targetName = mystrdup(argv[nextarg]);
    /* Else, let's build it with source's name with ASM extension. */
    else
        ira->filenames.targetName = ExtendFileName(ira->filenames.sourceName, ASM_EXT);

    /* Let's build other names from source name with appropriate extensions */
    ira->filenames.configName = ExtendFileName(ira->filenames.sourceName, CONFIG_EXT);
    ira->filenames.binaryName = ExtendFileName(ira->filenames.sourceName, BIN_EXT);
    ira->filenames.labelName = ExtendFileName(ira->filenames.sourceName, LABEL_EXT);

    /* If source type wasn't forced to binary, let's find out the file type. */
    if (!ira->params.sourceType)
        ira->params.sourceType = AutoScan(ira);

    /* If source type IS binary, of course, there won't be more than one Reloc */
    if (ira->params.sourceType == M68K_BINARY)
        ira->reloc.relocMax = 1;

    /* Let's try to get some memory */
    ira->label.labelAdr = mycalloc(ira->label.labelMax * 4);
    ira->reloc.relocAdr = mycalloc(ira->reloc.relocMax * 4);
    /* note: mycalloc() doesn't return if allocation failed, so no need to check returned value */
    ira->reloc.relocAdr[0] = 1; /* If no Reloc found */
    ira->reloc.relocOff = mycalloc(ira->reloc.relocMax * 4);
    ira->reloc.relocVal = mycalloc(ira->reloc.relocMax * 4);
    ira->reloc.relocMod = mycalloc(ira->reloc.relocMax * 4);
    ira->symbols.symbolName = mycalloc(ira->symbols.symbolMax * sizeof(char *));
    ira->symbols.symbolValue = mycalloc(ira->symbols.symbolMax * sizeof(uint32_t));
    ira->codeArea.codeArea1 = mycalloc(ira->codeArea.codeAreaMax * sizeof(uint32_t));
    ira->codeArea.codeArea2 = mycalloc(ira->codeArea.codeAreaMax * sizeof(uint32_t));
    ira->codeArea.cnfCodeArea1 = mycalloc(ira->codeArea.cnfCodeAreaMax * sizeof(uint32_t));
    ira->codeArea.cnfCodeArea2 = mycalloc(ira->codeArea.cnfCodeAreaMax * sizeof(uint32_t));
    ira->codeArea.codeAdr = mycalloc(ira->codeArea.codeAdrMax * sizeof(uint32_t));
    ira->noBase.noBaseStart = mycalloc(ira->noBase.noBaseMax * sizeof(uint32_t));
    ira->noBase.noBaseEnd = mycalloc(ira->noBase.noBaseMax * sizeof(uint32_t));
    ira->noPtr.noPtrStart = mycalloc(ira->noPtr.noPtrMax * sizeof(uint32_t));
    ira->noPtr.noPtrEnd = mycalloc(ira->noPtr.noPtrMax * sizeof(uint32_t));
    ira->text.textStart = mycalloc(ira->text.textMax * sizeof(uint32_t));
    ira->text.textEnd = mycalloc(ira->text.textMax * sizeof(uint32_t));
    ira->jmp.jmpTable = mycalloc(ira->jmp.jmpMax * sizeof(JMPTab_t));

    /* Source file read according to its chosen or detected type */
    switch (ira->params.sourceType & SOURCE_FAMILY_MASK) {
        case SOURCE_FAMILY_AMIGA:
            /* Let's try to open binary file */
            if ((ira->files.binaryFile = fopen(ira->filenames.binaryName, "wb")))
                if (ira->params.sourceType == AMIGA_HUNK_EXECUTABLE)
                    ReadAmigaHunkExecutable(ira);
                else
                    ReadAmigaHunkObject(ira);
            else
                ExitPrg("Can't open binary file \"%s\" for writing.", ira->filenames.binaryName);
            break;
        case SOURCE_FAMILY_ATARI:
            ReadAtariExecutable(ira);
            ExitPrg("Atari's executable files not yet supported.");
            break;
        case SOURCE_FAMILY_ELF:
            ReadElfExecutable(ira);
            ExitPrg("Elf's executable files not yet supported.");
            break;
        case SOURCE_FAMILY_NONE:
        default:
            Read68kBinary(ira);
            break;
    }

    /* Initial reading is done, let's close files to open them again differently, later */
    if (ira->files.sourceFile)
        fclose(ira->files.sourceFile);
    if (ira->files.binaryFile)
        fclose(ira->files.binaryFile);
    ira->files.binaryFile = ira->files.sourceFile = NULL;

    /* A binary file has been created, containing everything, its size is the program size */
    ira->params.prgLen = FileLength((uint8_t *)ira->filenames.binaryName);

    /* Open binary file */
    if (!(ira->files.binaryFile = fopen(ira->filenames.binaryName, "rb")))
        ExitPrg("Can't open binary file \"%s\" for reading.", ira->filenames.binaryName);

    /* Open labels file */
    if (!(ira->files.labelFile = fopen(ira->filenames.labelName, "wb")))
        ExitPrg("Can't open label file \"%s\" for writing.", ira->filenames.labelName);

    /* note: mycalloc() doesn't return if allocation failed, so no need to check returned value */
    ira->LabelNum = mycalloc(ira->hunkCount * sizeof(uint32_t));
    ira->XRefList = mycalloc(ira->LabX_len * sizeof(uint32_t));
    ira->buffer = mycalloc(ira->params.prgLen + 4);

    /* Has binary file the expected size ? */
    if ((fread(ira->buffer, 1, ira->params.prgLen, ira->files.binaryFile)) != ira->params.prgLen)
        ExitPrg("Can't read all data (binary file has not expected size).");

    /* Something obvious about program's end */
    ira->params.prgEnd = ira->params.prgStart + ira->params.prgLen;

    /* Now dealing with config file */
    if (ira->params.pFlags & CONFIG)
        ReadConfig(ira);

    /* Computing maximum length of a string containing the current address (between 1 and 8 because m68ks are 32bits wide) */
    if (ira->adrlen == 0)
        for (ira->adrlen = 1; (1 << (ira->adrlen * 4)) < ira->params.prgEnd && ira->adrlen < 8; ira->adrlen++)
            ;

    /* Reject code entry if too high, but override code entry to start of program if too low */
    if (ira->params.codeEntry >= ira->params.prgEnd)
        ExitPrg("ERROR: Entry(=$%08lX) is out of range!", (unsigned long) ira->params.codeEntry);
    if (ira->params.codeEntry < ira->params.prgStart)
        ira->params.codeEntry = ira->params.prgStart;

    if (ira->params.pFlags & BASEREG2) {
        /* Same as code entry with base address */
        if (ira->baseReg.baseAddress >= ira->params.prgEnd)
            ExitPrg("ERROR: BASEADR(=$%08lX) is out of range!", (unsigned long) ira->baseReg.baseAddress);
        if (ira->baseReg.baseAddress < ira->params.prgStart)
            ira->baseReg.baseAddress = ira->params.prgStart;

        InsertLabel(ira->baseReg.baseAddress);
        for (ira->baseReg.baseSection = 0; ira->baseReg.baseSection < ira->hunkCount; ira->baseReg.baseSection++)
            if (ira->baseReg.baseAddress >= ira->hunksOffs[ira->baseReg.baseSection] &&
                ira->baseReg.baseAddress < ira->hunksOffs[ira->baseReg.baseSection] + ira->hunksSize[ira->baseReg.baseSection])
                break;
    }

    printf("SOURCE : \"%s\"\n", ira->filenames.sourceName);
    printf("TARGET : \"%s\"\n", ira->filenames.targetName);
    if (ira->params.pFlags & KEEP_BINARY)
        printf("BINARY : \"%s\"\n", ira->filenames.binaryName);
    if (ira->params.pFlags & CONFIG)
        printf("CONFIG : \"%s\"\n", ira->filenames.configName);
    for (i = 0; i < cpuname_number; i++)
        if (ira->params.cpuType & (1 << i))
            printf("MACHINE: %s\n", cpuname[i]);
    printf("OFFSET : $%08lX\n", (unsigned long) ira->params.prgStart);
}

void ReadOptions(ira_t *ira, int argc, char **argv, int *nextarg, uint16_t *addrstyle) {
    char *odata, option, *data;

    while ((odata = argopt(argc, argv, nextarg, &option))) {
        switch (toupper(option)) {
            case 'E':
                if (!strnicmp(odata, "NTRY=", 5)) {
                    ira->params.codeEntry = parseAddress(&odata[5]);
                    if ((int32_t)(ira->params.codeEntry) < 0L)
                        ExitPrg("-ENTRY=%s: ENTRY must not be negative!", &odata[5]);
                } else if (!(stricmp(odata, "SCCODES")))
                    ira->params.pFlags |= ESCCODES;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'S':
                if (!(stricmp(odata, "PLITFILE")))
                    ira->params.pFlags |= SPLITFILE;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'P':
                if (!(stricmp(odata, "REPROC")))
                    ira->params.pFlags |= PREPROC;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'T':
                if (!(stricmp(odata, "EXT=1")))
                    ira->params.textMethod = 1;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'M':
                if (!strncmp(odata, "680", 3))
                    /* Processor stuff */
                    if (ira->params.cpuType & M680x0)
                        ExitPrg("Processor already defined.");
                    else if (!strcmp(&odata[2], "000"))
                        ira->params.cpuType |= M68000;
                    else if (!strcmp(&odata[2], "010"))
                        ira->params.cpuType |= M68010;
                    else if (!strcmp(&odata[2], "020"))
                        ira->params.cpuType |= M68020;
                    else if (!strcmp(&odata[2], "030"))
                        ira->params.cpuType |= M68030;
                    else if (!strcmp(&odata[2], "040"))
                        ira->params.cpuType |= M68040;
                    else if (!strcmp(&odata[2], "060"))
                        ira->params.cpuType |= M68060;
                    else
                        ExitPrg("Unknown processor Motorola %c%s", option, odata);
                else
                    /* MMU stuff */
                    if (!strcmp(odata, "68851"))
                    ira->params.cpuType |= M68851;
                else if (!strncmp(odata, "6888", 4))
                    /* FPU stuff */
                    if (ira->params.cpuType & FPU_EXTERNAL)
                        ExitPrg("Arithmetic coprocessor already defined.");
                    else if (!strcmp(odata, "68881"))
                        ira->params.cpuType |= M68881;
                    else if (!strcmp(odata, "68882"))
                        ira->params.cpuType |= M68882;
                    else
                        ExitPrg("Unknown arithmetic coprocessor Motorola %c%s", option, odata);
                else
                    ExitPrg("Unknown Motorola %c%s", option, odata);
                break;

            case 'A':
                ira->params.pFlags |= ADR_OUTPUT;
                if (!stricmp(odata, "W"))
                    ira->adrlen = 8;
                else if (odata[0])
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'O':
                if (!stricmp(odata, "LDSTYLE"))
                    if (*addrstyle == NEW_ADDR_STYLE)
                        ExitPrg("-OLDSTYLE cannot be use with -NEWSTYLE");
                    else
                        *addrstyle = OLD_ADDR_STYLE;
                else if (!strnicmp(odata, "FFSET=", 6)) {
                    ira->params.prgStart = parseAddress(&odata[6]);
                    if ((int32_t) ira->params.prgStart < 0L)
                        ExitPrg("-OFFSET=%s: OFFSET must not be negative!", &odata[6]);
                } else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'I':
                if (!(stricmp(odata, "NFO")))
                    ira->params.pFlags |= SHOW_RELOCINFO;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'C':
                if (!(stricmp(odata, "ONFIG")))
                    ira->params.pFlags |= CONFIG;
                else if (!(strnicmp(odata, "OMPAT=", 6))) {
                    char c, *p = odata + 6;

                    while ((c = *p++)) {
                        switch (tolower((unsigned) c)) {
                            case 'b':
                                ira->params.bitRange = 1;
                                break;
                            case 'i':
                                ira->params.immedByte = 1;
                                break;
                            default:
                                ExitPrg("Illegal COMPAT flag '%c'", c);
                                break;
                        }
                    }
                } else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'K':
                if (!(stricmp(odata, "EEPZH")))
                    ira->params.pFlags |= KEEP_ZEROHUNKS;
                else if (!(stricmp(odata, "EEPBIN")))
                    ira->params.pFlags |= KEEP_BINARY;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'N':
                if (!(stricmp(odata, "EWSTYLE")))
                    if (*addrstyle == OLD_ADDR_STYLE)
                        ExitPrg("-NEWSTYLE cannot be use with -OLDSTYLE");
                    else
                        *addrstyle = NEW_ADDR_STYLE;
                else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            case 'B':
                if (!(stricmp(odata, "INARY")))
                    ira->params.sourceType = M68K_BINARY;
                else if (!(stricmp(odata, "ASEABS")))
                    ira->params.baseAbs = 1;
                else if (!(stricmp(odata, "ASEREG")))
                    ira->params.pFlags |= BASEREG1;
                else if (!(strnicmp(odata, "ASEREG=", 7))) {
                    ira->params.baseReg = odata[7] - '0';
                    if ((data = strchr(odata, ','))) {
                        ira->baseReg.baseAddress = parseAddress(&data[1]);
                        ira->params.pFlags |= BASEREG2;
                        if ((data = strchr(&data[1], ','))) {
                            int32_t off;
                            off = stcd_l(&data[1]);
                            ira->baseReg.baseOffset = (int16_t) off;
                        } else
                            ira->baseReg.baseOffset = 0;
                    } else
                        ira->params.pFlags |= BASEREG1;
                    if (ira->params.baseReg > 7)
                        ExitPrg("Register cannot be greater than 7.");
                } else
                    ExitPrg("Unknown option -%c%s", option, odata);
                break;

            default:
                ExitPrg("Unknown option -%c%s", option, odata);
                break;
        }
    }
}

void CheckCPU(ira_t *ira) {
    /* If no processor have been specified on command line :
     * let's choose our plain old M68000 */
    if (!(ira->params.cpuType & M680x0))
        ira->params.cpuType |= M68000;

    /* MMU check :
     * only M68020 can have M68851
     * M68030, M68040 and M68060 already have an internal MMU
     * note : M68010 might have, in theory, by simulating its interface in software (but IRA won't care)
     * (http://en.wikipedia.org/wiki/Motorola_68851) */

    /* FPU check :
     * only M68020 and M68030 can have M68881 or M68882
     * M68040 and M68060 already have an internal FPU
     * note : 68881 and 68882 are functionally the same, they only differ by their inner structure and performances
     * (http://en.wikipedia.org/wiki/Motorola_68881) */

    /* If MMU selected and CPU is not 68020 */
    if ((ira->params.cpuType & M68851) && !(ira->params.cpuType & M68020))
        if (ira->params.cpuType & MMU_INTERNAL)
            ExitPrg("There is already a MMU in M68030, M68040 and M68060.");
        else
            ExitPrg("Only M68020 can have M68851.");
    else
        /* If FPU selected and CPU is neither 68020 nor 68030 */
        if ((ira->params.cpuType & FPU_EXTERNAL) && !(ira->params.cpuType & (M68020 | M68030))) {
            if (ira->params.cpuType & FPU_INTERNAL) {
                ExitPrg("There is already a FPU in M68040 and M68060.");
            } else {
                ExitPrg("Only M68020 and M68030 can have M68881 or M68882.");
            }
        }
}

char *ExtendFileName(char *filename, char *extension) {
    char *end = filename + strlen(filename);
    char *p = filename;
    char *new = myalloc(strlen(filename) + strlen(extension) + 1);
    char *node = new;

    while (*p && *p != ':')
        ++p;
    if (*p++ == ':')
        filename = p;
    p = end;
    while (p > filename && p[-1] != '/')
        --p;
    filename = p;
    p = end;
    while (p > filename && p[-1] != '.')
        --p;
    if (p > filename)
        end = p - 1;
    if (end > filename) {
        memcpy(node, filename, end - filename);
        node += end - filename;
    }
    *node = '\0';
    strcpy(node, extension);

    return new;
}
